PublishButton greyed out - MAGN 7569

holyjewsus opened the pull request:
… a can execute changed event on the submitCommand, this is raised when package contents changed when custom nodes or dlls are added


### Purpose
Can't upload packages, Publish! button is greyed out.
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7569

I could reproduce this bug when adding custom nodes or dlls as the last step in the package info form fillout, I.E adding name, description .... finally a custom node.  Customnodes raise property notification manually not in a setter, and when OnSubmit bool is checked again we do not raise a can execute changed event on the SubmitCommand, which is bound to the button state. I just added this invoke to the property change handler on the view model. 

![screen shot 2015-06-04 at 3 14 16 pm](https://cloud.githubusercontent.com/assets/508936/7992415/26bc00ae-0acd-11e5-9ac4-90f997369a4e.png)


### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [?] The level of testing this PR includes is appropriate, - @pboyer should this have a test? Not sure how this would work, is there a nice way to mock auth provider?

### Reviewers

@pboyer 
(if Peter does not have bandwidth, @ikeough or @lukechurch could merge)
### FYIs

@jnealb testing, have you encountered this bug when not adding package content as the last step?

jnealb commented:
@holyjewsus @pboyer The issue only occurs when adding the package as the last step.

pboyer commented:
Hi @holyjewsus,

Because this code was written before we adopted better dependency injection patterns, it's a bit tough to unit test this.  It can be done, though.  

You can mock nearly anything using the `Moq` library.  The pattern is typically:

```csharp
var mock = new Mock<Foo>();
mock.Setup( ... );
var mockObject = mock.Object;
```

You will see many examples of this in `PackageManagerClientTests` in DynamoCore.  

So, I would recommend going through the exercise of unit testing this behavior.

holyjewsus commented:
@ikeough @pboyer I've added a test that checks that the rCanExecuteChanged was raised on the submit command, but this test uses reflection, I'm not sure if thats okay.  I also noted that the package manager search tests continue to update the search results in the background during other tests, I'm not sure how to fix that, but I added a new sync context at the start of this test.

ikeough commented on src/DynamoCoreWpf/ViewModels/PackageManager/PublishPackageViewModel.cs:
```diff
@@ -507,7 +507,12 @@ public PublishPackageViewModel( DynamoViewModel dynamoViewModel ) : this()
+            if (e.PropertyName == "PackageContents")
+            {
+                CanSubmit();
+                BeginInvoke(
```
Why is BeginInvoke required here? Why can we not call RaiseCanExecuteChanged directly?

pboyer replied:
I agree with @ikeough.  It looks like you're using this method defensively.  I cannot conceive of a situation that this will be called from anything but the dispatcher thread to begin with.  This will also remove the nasty `DoEvents()` call in your test.

ikeough commented on test/DynamoCoreWpfTests/PackageManagerUITests.cs:
```diff
@@ -91,6 +95,39 @@ public void PackagePublishWindowClosesWithDynamo()
         }
 
+        [Test]
+        public void AdddingaPackageRaisesCanExecuteChangeOnDelegateCommand()
```
Spelling police - "AddingPackagesRaises..."

pboyer replied:
+1

ikeough commented on test/DynamoCoreWpfTests/PackageManagerUITests.cs:
```diff
@@ -91,6 +95,39 @@ public void PackagePublishWindowClosesWithDynamo()
+            var firstnode = customnodes.First();
+                      
+            var canExecuteChangedFired = 0;
+            vm.SubmitCommand.CanExecuteChanged += ((o, e) => { canExecuteChangedFired ++; });
```
If the BeginInvoke you use to set `CanExecuteChanged` is required, I worry that having a handler for `CanExecuteChanged` here won't be called while the test is still running (it's being invoked asynchronously). I know you use `DoEvents` later on, which I'm also not a big fan of because it's really difficult for the person reading this test to understand what is happening and in what order. It might be a necessary evil. Overall, this might cause some flaky behavior with this test.

ikeough commented:
@holyjewsus A couple of small comments here. In general, calling `RaiseCanExecuteChanged()` using `BeginInvoke` seems like overkill. I'm sure @pboyer had a good reason for it though, so we need to check with him to validate my comments here.

holyjewsus commented:
@pboyer can you comment on Ian's concerns? Should we be raising the notification directly or on the dispatcher?
is this test sufficiently ugly as to be useless?

pboyer commented on test/DynamoCoreWpfTests/PackageManagerUITests.cs:
```diff
@@ -91,6 +95,39 @@ public void PackagePublishWindowClosesWithDynamo()
+            //now use reflection to add a customnode to the package
+            var propertyChangedmethod = typeof(PublishPackageViewModel).GetMethods(BindingFlags.NonPublic | BindingFlags.Instance).Where(x=>x.Name == "AddFile").First();
+            propertyChangedmethod.Invoke(vm, BindingFlags.InvokeMethod,null, new object[] { firstnode },System.Globalization.CultureInfo.CurrentCulture);
+            
```
@holyjewsus If this is a private method, it's fine if you simply make it `internal` and call it here.  `internal` methods are exposed to DynamoCoreWpfTests using the `InternalsVisibleTo` property.

pboyer commented on test/DynamoCoreWpfTests/PackageManagerUITests.cs:
```diff
@@ -91,6 +95,39 @@ public void PackagePublishWindowClosesWithDynamo()
+            //running and throws exception when attempting to modify searchresults
+            SynchronizationContext.SetSynchronizationContext(new SynchronizationContext());
+            var vm = new PublishPackageViewModel(ViewModel);
+            ViewModel.OnRequestPackagePublishDialog(vm);
```
Let's not create a new instance.  Just reuse the one already available.

holyjewsus replied:
@pboyer to clarify: you mean use the current sync context, correct? If you mean the publishPackageViewModel, I'm not sure where one exits already?

pboyer commented:
@holyjewsus Please address the comments and make sure that you get the green light from appveyor.  Then you should be good to go.

holyjewsus commented:
@pboyer made the changes, appveyor shows green, tested again, seems good to go, just want to get your feedback on my last clarification above and then can merge.

holyjewsus commented:
@pboyer passes on CI, test failure was a failing test from Ben's fix for run settings on Home workspacemodel.  If you are okay with the test, then OK to merge.

pboyer commented:
@holyjewsus Can you merge this?
